rust/lance/src/io/writer/statistics.rs

```diff
@@ -150,12 +437,300 @@ fn get_boolean_statistics(arrays: &[&ArrayRef]) -> StatisticsRow {
     }
 
     StatisticsRow {
-        null_count,
+        null_count: ScalarValue::Int64(Some(null_count)),
         min_value: ScalarValue::Boolean(Some(true_present && !false_present)),
         max_value: ScalarValue::Boolean(Some(true_present || !false_present)),
     }
 }
 
+fn get_fixed_size_list_statistics<T: ArrowNumericType>(arrays: &[&ArrayRef]) -> StatisticsRow
+where
+    T::Native: Bounded + PartialOrd,
+    datafusion::scalar::ScalarValue: From<<T as ArrowPrimitiveType>::Native>,
+{
+    let DataType::FixedSizeList(f, len) = arrays[0].data_type() else {
+        todo!()
+    };
+    let len = std::cmp::min(*len as usize, 16); // TODO: is 16 a reasonable lenght?
+    let mut null_count: i64 = 0;
+    let mut max_value = vec![T::Native::min_value(); len].to_vec();
+    let mut min_value = vec![T::Native::max_value(); len].to_vec();
+
+    arrays
+        .iter()
+        .map(|x| x.as_fixed_size_list())
+        .for_each(|array| {
+            null_count += array.null_count() as i64;
+            for row in array.iter() {
+                if let Some(row) = row {
+                    let values = as_primitive_array::<T>(&row);
+                    for (i, val) in values.iter().enumerate() {
+                        if let Some(val) = val {
+                            if let Some(Ordering::Greater) = val.partial_cmp(&max_value[i]) {
+                                max_value[i] = val;
+                            }
+                            if let Some(Ordering::Less) = val.partial_cmp(&min_value[i]) {
+                                min_value[i] = val;
+                            }
+                        }
+                    }
+                } else {
+                    continue;
+                }
+            }
+        });
+
+    let min_value = min_value
+        .iter()
+        .map(|&x| ScalarValue::from(x))
+        .collect::<Vec<_>>();
+    let max_value = max_value
+        .iter()
+        .map(|&x| ScalarValue::from(x))
+        .collect::<Vec<_>>();
+    StatisticsRow {
+        null_count: ScalarValue::Int64(Some(null_count)),
+        min_value: ScalarValue::Fixedsizelist(Some(min_value), f.clone(), len as i32),
+        max_value: ScalarValue::Fixedsizelist(Some(max_value), f.clone(), len as i32),
+    }
+}
+
+fn get_list_statistics<T: ArrowNumericType, N>(arrays: &[&ArrayRef]) -> StatisticsRow
+where
+    T::Native: Bounded + PartialOrd,
+    datafusion::scalar::ScalarValue: From<<T as ArrowPrimitiveType>::Native>,
+    <T as ArrowPrimitiveType>::Native: OffsetSizeTrait,
+    N: arrow_array::OffsetSizeTrait,
+{
```

Comment:
For list statistics, I think we can just compute the statistics of the child array. That way we can optimize queries like:

```sql
list_has_any(x, [4, 5, 6])
```

If we compute for the child then we can just re-use the function for computing primitive stats to get the min and max.

But I'd also say this is much lower priority than doing better testing of the existing stats.

Reply:
I've refactored this to work directly on values and reuse `get_primitive_statistics`.